diff --git a/src/http/ngx_http_upstream.c b/src/http/ngx_http_upstream.c
index d29e503..1b58c85 100644
--- a/src/http/ngx_http_upstream.c
+++ b/src/http/ngx_http_upstream.c
@@ -6151,11 +6151,13 @@ ngx_http_upstream(ngx_conf_t *cf, ngx_command_t *cmd, void *dummy)
         return rv;
     }
 
+#if !NGX_HAVE_NACOS
     if (uscf->servers->nelts == 0) {
         ngx_conf_log_error(NGX_LOG_EMERG, cf, 0,
                            "no servers are inside upstream");
         return NGX_CONF_ERROR;
     }
+#endif
 
     return rv;
 }
diff --git a/src/os/unix/ngx_process_cycle.c b/src/os/unix/ngx_process_cycle.c
index 5bc5ce9..647a037 100644
--- a/src/os/unix/ngx_process_cycle.c
+++ b/src/os/unix/ngx_process_cycle.c
@@ -10,6 +10,9 @@
 #include <ngx_event.h>
 #include <ngx_channel.h>
 
+#if NGX_HAVE_AUXILIARY
+#include <ngx_auxiliary_module.h>
+#endif
 
 static void ngx_start_worker_processes(ngx_cycle_t *cycle, ngx_int_t n,
     ngx_int_t type);
@@ -130,6 +133,9 @@ ngx_master_process_cycle(ngx_cycle_t *cycle)
     ngx_start_worker_processes(cycle, ccf->worker_processes,
                                NGX_PROCESS_RESPAWN);
     ngx_start_cache_manager_processes(cycle, 0);
+#if NGX_HAVE_AUXILIARY
+    ngx_aux_start_auxiliary_processes(cycle, 0);
+#endif
 
     ngx_new_binary = 0;
     delay = 0;
@@ -215,6 +221,9 @@ ngx_master_process_cycle(ngx_cycle_t *cycle)
                 ngx_start_worker_processes(cycle, ccf->worker_processes,
                                            NGX_PROCESS_RESPAWN);
                 ngx_start_cache_manager_processes(cycle, 0);
+#if NGX_HAVE_AUXILIARY
+                ngx_aux_start_auxiliary_processes(cycle, 0);
+#endif
                 ngx_noaccepting = 0;
 
                 continue;
@@ -234,6 +243,9 @@ ngx_master_process_cycle(ngx_cycle_t *cycle)
             ngx_start_worker_processes(cycle, ccf->worker_processes,
                                        NGX_PROCESS_JUST_RESPAWN);
             ngx_start_cache_manager_processes(cycle, 1);
+#if NGX_HAVE_AUXILIARY
+            ngx_aux_start_auxiliary_processes(cycle, 1);
+#endif
 
             /* allow new processes to start */
             ngx_msleep(100);
@@ -248,6 +260,9 @@ ngx_master_process_cycle(ngx_cycle_t *cycle)
             ngx_start_worker_processes(cycle, ccf->worker_processes,
                                        NGX_PROCESS_RESPAWN);
             ngx_start_cache_manager_processes(cycle, 0);
+#if NGX_HAVE_AUXILIARY
+            ngx_aux_start_auxiliary_processes(cycle, 0);
+#endif
             live = 1;
         }
 
@@ -1189,3 +1204,10 @@ ngx_cache_loader_process_handler(ngx_event_t *ev)
 
     exit(0);
 }
+
+
+#if NGX_HAVE_AUXILIARY
+void ngx_worker_aux_process_init(ngx_cycle_t *cycle){
+  ngx_worker_process_init(cycle, -1);
+}
+#endif
diff --git a/src/os/unix/ngx_process_cycle.h b/src/os/unix/ngx_process_cycle.h
index 69495d5..6c0d3a3 100644
--- a/src/os/unix/ngx_process_cycle.h
+++ b/src/os/unix/ngx_process_cycle.h
@@ -58,4 +58,9 @@ extern sig_atomic_t    ngx_reopen;
 extern sig_atomic_t    ngx_change_binary;
 
 
+#if NGX_HAVE_AUXILIARY
+#define NGX_AUX_PATCHED 1
+void ngx_worker_aux_process_init(ngx_cycle_t *cycle);
+#endif
+
 #endif /* _NGX_PROCESS_CYCLE_H_INCLUDED_ */
