diff --git a/src/http/ngx_http_upstream.c b/src/http/ngx_http_upstream.c
index d04d91e..0802c29 100644
--- a/src/http/ngx_http_upstream.c
+++ b/src/http/ngx_http_upstream.c
@@ -6163,11 +6163,13 @@ ngx_http_upstream(ngx_conf_t *cf, ngx_command_t *cmd, void *dummy)
         return rv;
     }
 
+#if !NGX_HAVE_NACOS
     if (uscf->servers->nelts == 0) {
         ngx_conf_log_error(NGX_LOG_EMERG, cf, 0,
                            "no servers are inside upstream");
         return NGX_CONF_ERROR;
     }
+#endif
 
     return rv;
 }
diff --git a/src/os/unix/ngx_process_cycle.c b/src/os/unix/ngx_process_cycle.c
index cbe6dc7..cad0913 100644
--- a/src/os/unix/ngx_process_cycle.c
+++ b/src/os/unix/ngx_process_cycle.c
@@ -10,6 +10,9 @@
 #include <ngx_event.h>
 #include <ngx_channel.h>
 
+#if NGX_HAVE_AUXILIARY
+#include <ngx_auxiliary_module.h>
+#endif
 
 static void ngx_start_worker_processes(ngx_cycle_t *cycle, ngx_int_t n,
     ngx_int_t type);
@@ -135,6 +138,9 @@ ngx_master_process_cycle(ngx_cycle_t *cycle)
     ngx_start_worker_processes(cycle, ccf->worker_processes,
                                NGX_PROCESS_RESPAWN);
     ngx_start_cache_manager_processes(cycle, 0);
+#if NGX_HAVE_AUXILIARY
+    ngx_aux_start_auxiliary_processes(cycle, 0);
+#endif
     ngx_start_privileged_agent_processes(cycle, 0);
 
     ngx_new_binary = 0;
@@ -221,6 +227,9 @@ ngx_master_process_cycle(ngx_cycle_t *cycle)
                 ngx_start_worker_processes(cycle, ccf->worker_processes,
                                            NGX_PROCESS_RESPAWN);
                 ngx_start_cache_manager_processes(cycle, 0);
+#if NGX_HAVE_AUXILIARY
+                ngx_aux_start_auxiliary_processes(cycle, 0);
+#endif
                 ngx_start_privileged_agent_processes(cycle, 0);
                 ngx_noaccepting = 0;
 
@@ -241,6 +250,9 @@ ngx_master_process_cycle(ngx_cycle_t *cycle)
             ngx_start_worker_processes(cycle, ccf->worker_processes,
                                        NGX_PROCESS_JUST_RESPAWN);
             ngx_start_cache_manager_processes(cycle, 1);
+#if NGX_HAVE_AUXILIARY
+            ngx_aux_start_auxiliary_processes(cycle, 1);
+#endif
             ngx_start_privileged_agent_processes(cycle, 1);
 
             /* allow new processes to start */
@@ -256,6 +268,9 @@ ngx_master_process_cycle(ngx_cycle_t *cycle)
             ngx_start_worker_processes(cycle, ccf->worker_processes,
                                        NGX_PROCESS_RESPAWN);
             ngx_start_cache_manager_processes(cycle, 0);
+#if NGX_HAVE_AUXILIARY
+            ngx_aux_start_auxiliary_processes(cycle, 0);
+#endif
             ngx_start_privileged_agent_processes(cycle, 0);
             live = 1;
         }
@@ -1309,3 +1324,13 @@ ngx_cache_loader_process_handler(ngx_event_t *ev)
 
     exit(0);
 }
+
+
+#if NGX_HAVE_AUXILIARY
+void ngx_worker_aux_process_init(ngx_cycle_t *cycle) {
+  ngx_worker_process_init(cycle, -1);
+}
+void ngx_worker_aux_process_exit(ngx_cycle_t *cycle) {
+  ngx_worker_process_exit(cycle);
+}
+#endif
diff --git a/src/os/unix/ngx_process_cycle.h b/src/os/unix/ngx_process_cycle.h
index 5149396..e2776e0 100644
--- a/src/os/unix/ngx_process_cycle.h
+++ b/src/os/unix/ngx_process_cycle.h
@@ -59,4 +59,10 @@ extern sig_atomic_t    ngx_reopen;
 extern sig_atomic_t    ngx_change_binary;
 
 
+#if NGX_HAVE_AUXILIARY
+#define NGX_AUX_PATCHED 1
+void ngx_worker_aux_process_init(ngx_cycle_t *cycle);
+void ngx_worker_aux_process_exit(ngx_cycle_t *cycle);
+#endif
+
 #endif /* _NGX_PROCESS_CYCLE_H_INCLUDED_ */
